src/utp_stream.cpp

```diff
@@ -1360,6 +1375,9 @@ bool utp_socket_impl::send_pkt(int const flags)
 			m_fast_resend_seq_nr = (m_fast_resend_seq_nr + 1) & ACK_MASK;
 	}
 
+	if (state() == state_t::fin_sent && !force)
+		return false;
```

Comment:
is this the case where we attempt to send more, new, payload *after* having closed the connection?

Reply:
I think there might have been a particular case where this actually happened, though I may have also just fluffed this bit up. I originally did this a few weeks ago, and vaguely recall solving something with this, but can't remember exactly. I'll see if I can work out how (if!) I got this to happen. I imagine I might have triggered it while testing with 'real-world' torrents, rather than the simulations.

Reply:
```
        if (state() == state_t::fin_sent && !force)
	{
		UTP_LOGV("%8p: We can reach here\n"
			, static_cast<void*>(this));
		return false;
	}
```

It looks like it is possible for this to happen. It seems to come via `issue_write()`

```
[000325524325] 0x7f806a583d10: skipping send (no payload and no force) seq_nr:56054 ack_nr:47142 id:39595 target:xxx header_size:20 error:No error information send_buffer_size:0 cwnd:3634 adv_wnd:1048576 in-flight:238 mtu:1241
[000325524334] 0x7f806a583d10: defer ack
[000325524334] 0x7f806a583d10: subscribe drained
[000325524335] 0x7f806a583d10: incoming stream consumed
[000325524338] 0x7f806a583d10: sending packet seq_nr:56054 ack_nr:47142 type:ST_FIN id:39595 target:xxx size:20 error:No error information send_buffer_size:0 cwnd:3634 adv_wnd:1048576 in-flight:238 mtu:1241 timestamp:1462880446 time_diff:185335 mtu_probe:0 extension:0
[000325524373] 0x7f806a583d10: Cancelling redundant deferred ack
[000325524374] 0x7f806a583d10: state:FIN_SENT
[000325524572] 0x7f806a583d10: calling read handler read:41843
[000325524574] 0x7f806a583d10: calling read handler read:41843 ec:No error information shutdown:0
[000325524643] 0x7f806a583d10: add_read_buffer 327502 bytes
[000325525006] 0x7f806a583d10: Didn't fill entire target: 0 bytes left in buffer
[000325525008] 0x7f806a583d10: 240 packets moved from buffer to user space (327502 bytes)
[000325525830] 0x7f806a583d10: add_read_buffer 377133 bytes
[000325525832] 0x7f806a583d10: new read handler. 0 bytes in buffer
[000325525837] 0x7f806a583d10: add_write_buffer 128 bytes
[000325525838] 0x7f806a583d10: add_write_buffer 128 bytes
[000325525839] 0x7f806a583d10: add_write_buffer 128 bytes
[000325525840] 0x7f806a583d10: add_write_buffer 128 bytes
[000325525840] 0x7f806a583d10: add_write_buffer 128 bytes
[000325525841] 0x7f806a583d10: add_write_buffer 128 bytes
[000325525842] 0x7f806a583d10: add_write_buffer 14 bytes
[000325525843] 0x7f806a583d10: new write handler. 782 bytes to write
[000325525844] 0x7f806a583d10: We can reach here
```

Reply:
interesting. trying to write to a socket that we have closed should probably fail earlier, long before we add the data to the send buffer even. It also indicates that something fishy is going on a layer up. There shouldn't be any new data written to sockets after they're closed.

Reply:
Triggering more data being sent while the connection is closing seems to be a bit of a special case. Picking up existing write buffer or nagled packets when sending acks or the fin itself should already be prevented (`finalizing == true`), although in this case its when more data is being added to the write buffer. This seems to only sometimes happen.

My current understanding is that the other end needs to be the first to send the fin for this to occur. Maybe in this case the BT layer isn't aware that the connection is closing until it's actually deleted. Is `cancel_handlers()` when it finds out? Could this maybe be called earlier when the other end is the first to send a fin?

Reply:
consider the case where we have a bunch of packets in-flight. Before those packets are ACKed, we close the socket and send a FIN.

Some of those outstanding packets might need to be re-sent, which would take this path, even when we're in `fin_sent` mode. I think this invariant really should be to not send any packets with sequence numbers past `m_seq_nr`

Reply:
oh, I see. This is *after* the loop that's re-sending packets.